compiler/compiler_test.go

```diff
@@ -41,27 +43,336 @@ func Bfunc() int {
 	return varA+varB
 }
 `
+
 	files := []source{{"fileA.go", []byte(fileA)}, {"fileB.go", []byte(fileB)}}
 
-	compare(t, "foo", files, false)
-	compare(t, "foo", files, true)
+	compareOrder(t, files, false)
+	compareOrder(t, files, true)
 }
 
-func compare(t *testing.T, path string, sourceFiles []source, minify bool) {
-	outputNormal, err := compile(path, sourceFiles, minify)
-	if err != nil {
-		t.Fatal(err)
-	}
+func TestDeclSelection_KeepUnusedExportedMethods(t *testing.T) {
+	src := `
+		package main
+		type Foo struct {}
+		func (f Foo) Bar() {
+			println("bar")
+		}
+		func (f Foo) Baz() { // unused
+			println("baz")
+		}
+		func main() {
+			Foo{}.Bar()
+		}`
+
+	srcFiles := []source{{`main.go`, []byte(src)}}
+	sel := declSelection(t, srcFiles, nil)
+
+	sel.DeclCode.IsAlive(`^\s*Foo = \$newType`)
```

Comment:
Depending on the generated code is somewhat fragile. Maybe we could use `Decl.FullName` instead, at least for decls that support it?

Reply:
Many of the `Decl.FullNames` are blank or the generic name duplicated on each instance. I totally agree, it seems fragile to use the generated code but it's the only thing that I could used to differentiate the decls sometimes. 🤷‍♂️ 

I tried to make it less fragile by using regular expressions so that code like `Bar[42 /* int */]` skips checking the number and only checks the "Bar" and "int" part, i.e. `Bar\[\d+ /\* int \*/\]`. It also made it easy to jump over those position marks (`\b`) in the code. But I realize that's not perfect. I thought about making methods like `IsAliveInstance("Bar", "int")` so that if it did change we'd have less to update, but it obscured what was being checked.

I created `PrintDeclStatus()` as one of the first things to see what Decls were in the package and what information they had. (Running debug wasn't very helpful to me since those were `[]byte`s) It prints the `Decl.FullNames` on the same line as `[Alive]` or `[Dead]`. The following is the output from adding `PrintDeclStatus` to `TestDeclSelection_RemoveUnusedStructTypeInstances`:

```Plain
Package command-line-arguments
  [Alive] ""
     DeclCode: "\tFoo = {};\n"
     TypeInitCode: "\t$pkg.Foo = Foo;\n"
     Vars: [Foo]
  [Alive] ""
     DeclCode: "\tFoo[0 /* float64 */] = $newType(0, $kindStruct, \"main.Foo[float64]\", true, \"command-line-arguments\", true, function(v_) {\n\t\tthis.$val = this;\n\t\tif (arguments.length === 0) {\n\t\t\tthis.v = 0;\n\t\t\treturn;\n\t\t}\n\t\tthis.v = v_;\n\t});\n"
     MethodListCode: "\tFoo[0 /* float64 */].methods = [{prop: \"Bar\", name: \"Bar\", pkg: \"\", typ: $funcType([], [], false)}];\n"
     TypeInitCode: "\tFoo[0 /* float64 */].init(\"command-line-arguments\", [{prop: \"v\", name: \"v\", embedded: false, exported: false, typ: $Float64, tag: \"\"}]);\n"
  [Alive] ""
     DeclCode: "\tFoo[2 /* int */] = $newType(0, $kindStruct, \"main.Foo[int]\", true, \"command-line-arguments\", true, function(v_) {\n\t\tthis.$val = this;\n\t\tif (arguments.length === 0) {\n\t\t\tthis.v = 0;\n\t\t\treturn;\n\t\t}\n\t\tthis.v = v_;\n\t});\n"
     MethodListCode: "\tFoo[2 /* int */].methods = [{prop: \"Bar\", name: \"Bar\", pkg: \"\", typ: $funcType([], [], false)}];\n"
     TypeInitCode: "\tFoo[2 /* int */].init(\"command-line-arguments\", [{prop: \"v\", name: \"v\", embedded: false, exported: false, typ: $Int, tag: \"\"}]);\n"
  [Dead]  ""
     InitCode: "\b\x00\x00\x00i\t\t$unused(new Foo[0 /* float64 */].ptr(3.14));\n"
     Vars: [_]
  [Alive] "(command-line-arguments.Foo[T]).Bar"
     DeclCode: "\t$ptrType(Foo[0 /* float64 */]).prototype.Bar = function Foo·Bar() {\n\t\tvar f;\n\t\tf = this;\n\b\x00\x00\x00O\t\tconsole.log(f.v);\n\b\x00\x00\x00\x00\t};\n\tFoo[0 /* float64 */].prototype.Bar = function(...$args) { return this.$val.Bar(...$args); };\n"
  [Alive] "(command-line-arguments.Foo[T]).Bar"
     DeclCode: "\t$ptrType(Foo[2 /* int */]).prototype.Bar = function Foo·Bar$1() {\n\t\tvar f;\n\t\tf = this;\n\b\x00\x00\x00O\t\tconsole.log(f.v);\n\b\x00\x00\x00\x00\t};\n\tFoo[2 /* int */].prototype.Bar = function(...$args) { return this.$val.Bar(...$args); };\n"
  [Alive] ""
     Vars: [main]
  [Alive] "command-line-arguments.main"
     DeclCode: "\tmain = function main$1() {\n\b\x00\x00\x00\xa1\t\t$clone(new Foo[2 /* int */].ptr(7), Foo[2 /* int */]).Bar();\n\b\x00\x00\x00\x00\t};\n"
  [Alive] ""
     InitCode: "\b\x00\x00\x00\x00\t\tif ($pkg === $mainPkg) {\n\b\x00\x00\x00\x95\t\t\tmain();\n\b\x00\x00\x00\x00\t\t\t$mainFinished = true;\n\b\x00\x00\x00\x00\t\t}\n"
```

Reply:
I see, so a principled fix would be to go and generate sensible identifiers for each kind of decl we emit, and use them in the test. I realize that could be tricky in the general case because it reduces to the problem of unique type names we discussed in the chat.

We could consider generating identifiers that are "good enough" to distinguish decls in our test cases, but I'll let you decide if you want to do that or not.

Reply:
@nevkontakte, since I'm not sure what it would take to name those (like you said about the sensible identifiers) and since this is a dependency for #1340, I'd prefer to get this in sooner and do a naming decls / stability upgrade as a later effort.

_(And, since I have managers that track the Jira tickets I make for these PRs, I know they'd like to see this progressing, even if it means I open a bunch of follow up Jira tickets. Besides they understand we're on your schedule not ours, and I care more about helping out with good code then about what they want. 😁 Hi, Dave, if you're seeing this 😆)_

Reply:
No worries, like I said, this was an optional suggestion.